src/operator.jl

```diff
@@ -13,6 +16,131 @@ for f in (:fragtype_a, :fragtype_b, :fragtype_accum, :load_a, :load_b, :load_c,
+
+        @inline function load_b(::Type{<:GeneralFPUOp{M, N, K, DT, CT}}, ::Type{$layout_type{CT}}, workspace, tile::Tile) where {M, N, K, DT, CT}
+            laneId = (threadIdx().x - 1) % 32 + 1
+
+            op_x = (laneId - 1) ÷ 4 + 1
+            y, x = (tile.base.K + tile.offset.K + 1, tile.base.N + tile.offset.N + op_x)
+
+            frag = LocalArray{Tuple{K, N ÷ 8}, CT}(undef)
+            @unroll for n = 1 : N ÷ 8
+                @unroll for k = 1 : K
+                    y_layout, x_layout = $convert_index_func((y + (k - 1), x + 8 * (n - 1)))
+                    @inbounds frag = setindex(frag, CT(workspace[y_layout, x_layout]), k, n)
+                end
+            end
+
+            return NTuple{K * N ÷ 8, CT}(frag)
+        end
+
+        @inline function load_c(::Type{<:GeneralFPUOp{M, N, K, DT, CT}}, ::Type{$layout_type{DT}}, workspace, tile::Tile) where {M, N, K, DT, CT}
+            laneId = (threadIdx().x - 1) % 32 + 1
+
+            op_y = (laneId - 1) % 4 + 1
+            op_x = (laneId - 1) ÷ 4 + 1
+
+            y, x = (tile.base.M + tile.offset.M + op_y, tile.base.N + tile.offset.N + op_x)
+
+            frag = LocalArray{Tuple{M ÷ 4, N ÷ 8}, DT}(undef)
+            @unroll for m = 1 : M ÷ 4
+                @unroll for n = 1 : N ÷ 8 
+                    @inbounds frag = setindex(frag, DT(workspace[y + 4 * (m - 1), x + 8 * (n - 1)]), m, n)
+                end
+            end
+
+            return NTuple{M * N ÷ 32, DT}(frag)
+        end
+
+        @inline function store_d(::Type{<:GeneralFPUOp{M, N, K, DT, CT}}, ::Type{$layout_type{DT}}, workspace, frag, tile::Tile) where {M, N, K, DT, CT}
+            laneId = (threadIdx().x - 1) % 32 + 1
+
+            op_y = (laneId - 1) % 4 + 1
+            op_x = (laneId - 1) ÷ 4 + 1
+
+            y, x = (tile.base.M + tile.offset.M + op_y, tile.base.N + tile.offset.N + op_x)
+
+            frag = LocalArray{Tuple{M ÷ 4, N ÷ 8}, DT}(frag)
+            @unroll for m = 1 : M ÷ 4
+                @unroll for n = 1 : N ÷ 8 
+                    @inbounds workspace[y + 4 * (m - 1), x + 8 * (n - 1)] = frag[m, n]
+                end
+            end
+        end
+    end
+end
+
+abstract type FPUOp{M, N, K, DT, CT} <: GeneralFPUOp{M, N, K, DT, CT} end
+function operator_fma(::Type{FPUOp{M, N, K, DT, CT}}, a::CT, b::CT, c::DT) where {M, N, K, DT, CT}
+    return fma(a, b, c)
+end
+
+abstract type TropicalFPUOp{M, N, K, DT, CT} <: GeneralFPUOp{M, N, K, DT, CT} end
+function operator_fma(::Type{TropicalFPUOp{M, N, K, DT, CT}}, a::CT, b::CT, c::DT) where {M, N, K, DT, CT}
+    return max(a + b, c)
+end
+
+for (operator_type) in [FPUOp, TropicalFPUOp]
+    @eval begin
+        @inline function mma(::Type{$operator_type{M, N, K, DT, CT}}, a_frag, b_frag, c_frag) where {M, N, K, DT, CT}
+            a_frag = LocalArray{Tuple{M ÷ 4, K}, CT}(a_frag)
+            b_frag = LocalArray{Tuple{K, N ÷ 8}, CT}(b_frag)
+            c_frag = LocalArray{Tuple{M ÷ 4, N ÷ 8}, DT}(c_frag)
```

Comment:
Is the casting necessary here? That is to say, can't you just return 2D arrays in `load_a`, `load_b`, and `load_c` in the first place?

Reply:
No, that was actually one of the major pain points in this implementation. Returning a `LocalArray` results in a segmentation fault and crashes the Julia REPL. You can verify this by just changing the return type for the A fragment, returning frag in `load_a` and commenting out the first line of this snippet. Casting to a `Tuple` was the workaround I managed to find to mitigate this issue.